build(fluid-runner): use node16 module resolution

sonalideshpandemsft opened the pull request:
Additions:
1. Update `parseBundleAndExportFile.ts` to use import 
```ts
const codeLoaderBundle = await import(codeLoader);
```
2. Upgrade `yargs` to `17.7.2` and `@types/yargs` to `17.0.32`. As of v16, yargs supports ESM imports
3. Add sample executable scripts in both ESM and CJS formats
4. Add tsconfig.bin.lint.json file to extend eslint coverage to the bin folder

Original PR: https://github.com/microsoft/FluidFramework/pull/20245

DLehenbauer committed:
build(fluid-runner): use node16 module resolution

RishhiB committed:
export IExportFileResponseSuccess, IExportFileResponseFailure, IFileLogger

DLehenbauer committed:
fix __dirname

RishhiB committed:
failing when importing into mocharc

RishhiB committed:
Revert "failing when importing into mocharc"

This reverts commit 2aef1844200bd672a62fb304c5e4e305f8ba3616.

DLehenbauer committed:
fix bin imports

DLehenbauer committed:
Merge branch 'fluid-runner' of https://github.com/microsoft/FluidFramework into fluid-runner

DLehenbauer committed:
skip esm tests

DLehenbauer committed:
remove unused api-extractor warning suppressions

DLehenbauer committed:
revert mocha-test-setup.api.md

DLehenbauer committed:
restore cc

sonalideshpandemsft committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into fluid-runner

RishhiB committed:
Merge branch 'main' into fluid-runner

sonalideshpandemsft committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into sonali/fluid-runner

sonalideshpandemsft committed:
cjs

sonalideshpandemsft committed:
remove require

sonalideshpandemsft committed:
esm testing

sonalideshpandemsft committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into sonali/fluid-runner

jason-ha commented on packages/tools/fluid-runner/src/index.ts:
```diff
@@ -4,15 +4,21 @@
+	createContainerAndExecute,
+	exportFile,
+	IExportFileResponse,
+	IExportFileResponseSuccess,
+	IExportFileResponseFailure,
+} from "./exportFile.js";
+export { fluidRunner } from "./fluidRunner.js";
+export { OutputFormat, ITelemetryOptions, IFileLogger } from "./logger/fileLogger.js";
```
I see exports were added to resolve api-extractor issues. Do we know the history for the prior choices? ...to go against api-extractor and not export those? Such changes are best done explicitly and separately.

sonalideshpandemsft replied:
api-extractor forgotten-export warning were suppressed before: https://github.com/microsoft/FluidFramework/pull/20349/files#diff-04ef5f87a3b831c02ccb64223b9bfc0bb33d0403d8b2f250179dbaf9e50267f3

jason-ha reviewed (changes requested):
Description should be updated to note the ESM support.

sonalideshpandemsft committed:
Merge remote-tracking branch 'origin' into sonali/fluid-runner

sonalideshpandemsft committed:
apply new lint rule

sonalideshpandemsft committed:
upgrade yargs to 17

sonalideshpandemsft commented on packages/tools/fluid-runner/bin/fluid-runner.cjs:
```diff
@@ -0,0 +1,6 @@
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+require("../dist/fluidRunner.js").fluidRunner();
```
Converting this to ESM requires some update to fluidRunner. As of v16, yargs supports ESM imports

sonalideshpandemsft commented on packages/tools/fluid-runner/src/test/fluidRunner.spec.ts:
```diff
@@ -192,11 +193,11 @@ describe("fluid-runner from command line", () => {
 
 describe("custom fluidFileConverter provided", () => {
-	const command = path.join(__dirname, "../../src/test/sampleCodeLoaders", "sample-executable");
+	const command = path.join(_dirname, "../../bin/sample-executable.cjs");
```
The move was made just to have all scripts under bin (organizational purpose only)

sonalideshpandemsft committed:
mv back sample-executable

sonalideshpandemsft committed:
revert runtime typetests

jason-ha commented on packages/tools/fluid-runner/bin/fluid-runner.mjs:
Does .mjs extension by us something? Or was there a benefit to having no extension before?
From what I know .mjs is useful when it needs interpreted as ESM, but this package has "type": "module"; so .js or no extension should already be expected to be ESM. (I tend to assume someone had a reason for no extension, but I don't know what it is.) I do like the explicitness of .mjs; so, no need to change.

sonalideshpandemsft replied:
afair, mocha test failures on the old PR suggested to add extension (js/mjs/cjs) to this script: https://dev.azure.com/fluidframework/public/_build/results?buildId=250799&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&t=5cfe65a4-d7cb-5934-e953-1c67617821f4

```ts
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension "" for /mnt/vss/_work/1/s/packages/tools/fluid-runner/bin/fluid-runner. Loading extensionless files is not supported inside of "type":"module" package.json contexts. The package.json file /mnt/vss/_work/1/s/packages/tools/fluid-runner/package.json caused this "type":"module" context. Try changing /mnt/vss/_work/1/s/packages/tools/fluid-runner/bin/fluid-runner to have a file extension. Note the "bin" field of package.json can point to a file with an extension, for example {"type":"module","bin":{"fluid-runner":"./bin/fluid-runner.js"}}
packages/tools/fluid-runner test:mocha:     at new NodeError (node:internal/errors:405:5)
packages/tools/fluid-runner test:mocha:     at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:79:11)
packages/tools/fluid-runner test:mocha:     at defaultGetFormat (node:internal/modules/esm/get_format:124:36)
packages/tools/fluid-runner test:mocha:     at defaultLoad (node:internal/modules/esm/load:84:20)
packages/tools/fluid-runner test:mocha:     at nextLoad (node:internal/modules/esm/loader:163:28)
packages/tools/fluid-runner test:mocha:     at ESMLoader.load (node:internal/modules/esm/loader:603:26)
packages/tools/fluid-runner test:mocha:     at ESMLoader.moduleProvider (node:internal/modules/esm/loader:457:22)
packages/tools/fluid-runner test:mocha:     at new ModuleJob (node:internal/modules/esm/module_job:64:26)
packages/tools/fluid-runner test:mocha:     at #createModuleJob (node:internal/modules/esm/loader:480:17)
packages/tools/fluid-runner test:mocha:     at ESMLoader.getModuleJob (node:internal/modules/esm/loader:434:34) {
packages/tools/fluid-runner test:mocha:   code: 'ERR_UNKNOWN_FILE_EXTENSION'
```

jason-ha commented on packages/tools/fluid-runner/src/test/sampleCodeLoaders/sample-executable.mjs:
This means missing testing of this for CJS, which we should have if we are keeping CJS support. I think I have a kludge for how to test both.

jason-ha reviewed (changes requested):
I'll try to push some lint and correctness fixes Tuesday.

jason-ha committed:
fix(fluid-runner): use on Windows with absolute paths
update to use dynamic `import()` requires that absolute path have a file:// prefix on Windows to distinguish any drive letter from a protocol.

jason-ha committed:
fix(fluid-runner): lint/test coverage and corrections
- fluidRunner should not have floating (unobserved) Promise and has been made async with a Promise returned to account for that.
- Linting was added to all files including bin/fluid-runner.mjs which prompted further change.
- Cleanup unused lint disables
- Refactor test with sample-executable to cover CJS and ESM

sonalideshpandemsft commented on packages/tools/fluid-runner/tsconfig.bin.json:
```diff
@@ -0,0 +1,9 @@
+{
+	"extends": "../../../common/build/build-common/tsconfig.node16.json",
+	"include": ["bin/*"],
+	"references": [
```
why is tsconfig.bin.json required?

jason-ha replied:
We need some TS project file to provide coverage with eslint. Probably should rename it to have .`lint` in the name like was done for CJS test case.

sonalideshpandemsft commented on packages/tools/fluid-runner/src/test/sampleCodeLoaders/sample-executable.cjs.ts:
```diff
@@ -0,0 +1,12 @@
+ * Licensed under the MIT License.
+ */
+
+import { fluidRunner } from "../../index.js";
```
why import differently? why not the below? 
```ts
import { fluidRunner } from "@fluidframework/fluid-runner";
```

jason-ha replied:
"@fluidframework/fluid-runner" is the nice way to do it. You could take this example and use it outside of this package.

In the CJS case (surprise to me) the "self-reference" was an error according to VS Code's tsc result. I didn't think that support mattered CJS/ESM and just confirmed it looks like a VS Code limitation/bug. tsc is happy with @fluidframework/fluid-runner. I will update. Thanks for asking. :)

jason-ha replied:
Well, I don't know how VS Code is magically right about what should be done. Comment added.

sonalideshpandemsft commented on packages/tools/fluid-runner/src/test/fluidRunner.spec.ts:
```diff
@@ -193,11 +193,13 @@ describe("fluid-runner from command line", () => {
-		"../../src/test/sampleCodeLoaders",
-		"sample-executable.mjs",
+		"sampleCodeLoaders",
+		_dirname.includes("dist") ? "sample-executable.cjs.js" : "sample-executable.esm.js",
```
love this change. let's us test our script with both ESM and CJS :)

jason-ha committed:
test(fluid-runner): consistency in names and imports

sonalideshpandemsft committed:
resolve merge conflicts

jason-ha committed:
style(fluid-runner): import-lint formating

jason-ha committed:
test(fluid-runner): fix dual-emit import limitation

jason-ha committed:
test(fluid-runner): fix/suppress policy violations
